tests/jvm-log4j-backend-our-context/src/test/kotlin/Log4jRecorder.kt

```diff
@@ -0,0 +1,200 @@
+import org.apache.logging.log4j.Level as L4jLevel
+
+/**
+ * Records logging events produced by Log4J loggers.
+ */
+internal class Log4jRecorder(loggerName: String, minLevel: Level): Recorder(minLevel) {
+
+    private val logger: Logger = (LogManager.getLogger(loggerName) as Logger)
+    private var prevRootLevel: L4jLevel? = null
+    private var prevLevel: L4jLevel? = null
+    private var appender: Appender? = null
+
+    override fun start() {
+        logger.apply {
+            prevLevel = level
+            level = minLevel.toLog4j()
+        }
+        appender = object: AbstractAppender(logger.name, null, null, true, null) {
+            override fun append(event: LogEvent) {
+                val data = Log4jLogData(event)
+                append(data)
+            }
+        }
+        logger.addAppender(appender)
+
+        // Set the root logger level to `INFO` to be at the same level as `java.util.logging`.
+        // This is the assumption of the common tests.
+        val rootLogger = LogManager.getRootLogger() as Logger
+        prevRootLevel = rootLogger.level
+        rootLogger.level = INFO.toLog4j()
+    }
+
+    override fun stop() {
+        prevLevel?.let {
+            logger.level = it
+            prevLevel = null
+        }
+        appender?.let {
+            logger.removeAppender(it)
+            appender = null
+        }
+        prevRootLevel?.let {
+            (LogManager.getRootLogger() as Logger).level = prevRootLevel
+            prevRootLevel = null
+        }
+    }
+}
+
+/**
+ * Adapts [LogEvent] to [LogData] used by [Recorder].
+ */
+private class Log4jLogData(val event: LogEvent): LogData {
+    override val level: Level by lazy {
+        event.level.toLevel()
+    }
+    override val message: String = event.message.formattedMessage
+    override val throwable: Throwable? = event.thrown
+    override fun equals(other: Any?): Boolean {
+        return if (other is Log4jLogData) {
+            event == other.event
+        } else {
+            false
+        }
+    }
+    override fun hashCode(): Int = event.hashCode()
+}
+
+/**
+ * A reasonably high value to be used for the range calculations instead of
+ * [Int.MAX_VALUE] or [Int.MIN_VALUE] used as high/low bounds in logging frameworks.
+ *
+ * This is a safe assumption that log levels are defined withing this range,
+ * especially for tests.
+ */
+private const val BOUND = 1200 // java.util.logging.Level.intValue() + 1000
+
+private val julProgression: IntProgression = BOUND.downTo(0)
+
+private fun IntProgression.size(): Int = (last - first).absoluteValue + 1
+
+private val log4jRange: IntRange = 0.rangeTo(BOUND)
+
+private fun IntRange.size(): Int = (last - first + 1)
+
+@Suppress("unused") // Is meant to force the construction of the converter object.
+private val forceRegistration = Log4jLevelConverter
+
+/**
+ * Converts this logging level instance to a Log4j counterpart.
+ */ //VisibleForTesting
```

Comment:
@alexander-yevsyukov 
Could you please do something with this comment? Let's either move it inside the comment, or use a `Suppress`.